File: trackers/trax.py
from datetime import datetime
from .base import BaseTracker


class TraxTracker(BaseTracker):
    """Tracker for Trax courier parcels."""

    API_URL = "https://sonic.pk/api/shipment/track/consignee/public"

    def track(self, tracking_number):
        """Track a Trax parcel using their public API."""
        try:
            resp = self.session.get(
                self.API_URL,
                params={'tracking_number': tracking_number},
                timeout=30
            )
            resp.raise_for_status()
            data = resp.json()

            if data.get('status') != 1 or not data.get('details'):
                msg = data.get('message', 'Not found')
                return {"error": f"Tracking not found: {msg}"}

            # Get the first (and only) shipment detail
            detail = next(iter(data['details'].values()))

            # Current status from first history entry
            history = detail.get('tracking_history', [])
            status = history[0]['status'] if history else 'Unknown'

            # Location
            consignee = detail.get('consignee', {})
            location = consignee.get('destination', 'N/A')

            # Build history list
            history_list = []
            for h in history:
                history_list.append({
                    'timestamp': h.get('date_time', ''),
                    'status': h.get('status', ''),
                    'location': ''
                })

            delivered = 'delivered' in status.lower()

            return {
                'status': status,
                'location': location,
                'timestamp': datetime.now().isoformat(),
                'history': history_list,
                'delivered': delivered
            }

        except Exception as e:
            return {"error": f"Tracking failed: {str(e)}"}

File: src/server/trackers/trax.ts
import { TrackingResult, HistoryItem } from './types.js';

export async function trackTrax(trackingNumber: string): Promise<TrackingResult> {
  const url = `https://sonic.pk/api/shipment/track/consignee/public?tracking_number=${encodeURIComponent(trackingNumber)}`;
  try {
    const controller = new AbortController();
    const timeout = setTimeout(() => controller.abort(), 20000);

    const resp = await fetch(url, {
      method: "GET",
      headers: {
        "Accept": "application/json, text/plain, */*",
        "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36"
      },
      signal: controller.signal
    });
    clearTimeout(timeout);

    if (!resp.ok) {
      return {
        status: "Error",
        location: "N/A",
        timestamp: new Date().toISOString(),
        history: [],
        delivered: false,
        error: `HTTP ${resp.status} ${resp.statusText}`
      };
    }

    const data = await resp.json() as any;

    if (data?.status !== 1 || !data?.details) {
      const msg = data?.message || "Not found";
      return {
        status: "Error",
        location: "N/A",
        timestamp: new Date().toISOString(),
        history: [],
        delivered: false,
        error: `Tracking not found: ${msg}`
      };
    }

    const detailsValues = Object.values(data.details);
    if (detailsValues.length === 0) {
      return {
        status: "Error",
        location: "N/A",
        timestamp: new Date().toISOString(),
        history: [],
        delivered: false,
        error: "No shipment details returned"
      };
    }

    const detail: any = detailsValues[0];
    const rawHistory = Array.isArray(detail?.tracking_history) ? detail.tracking_history : [];
    const status = rawHistory.length > 0 ? (rawHistory[0].status || "Unknown") : "Unknown";

    const consignee = detail?.consignee || {};
    const location = consignee.destination || "N/A";

    const history: HistoryItem[] = [];
    for (const h of rawHistory) {
      history.push({
        timestamp: h.date_time || "",
        status: h.status || "",
        location: ""
      });
    }

    const delivered = status.toLowerCase().includes("delivered");

    return {
      status,
      location,
      timestamp: new Date().toISOString(),
      history,
      delivered
    };
  } catch (err: any) {
    return {
      status: "Error",
      location: "N/A",
      timestamp: new Date().toISOString(),
      history: [],
      delivered: false,
      error: `Tracking failed: ${err.message || String(err)}`
    };
  }
}
